chore(Notification): Convert dismiss ambient notification dialog

geoffreykwan opened the pull request:
## Changes

Converted notificationDismissDialog.html to an Angular component.

## Test

You can use the component content JSON below to test the ambient notification. Have the student submit and get a score of 1 three times in a row and it should trigger the ambient notification. Then you can test the dismiss dialog.

- Test the dialog when there is a dismiss code
- Test the dialog when there is no dismiss code

```
{
    "id": "0k0r4jlona",
    "type": "OpenResponse",
    "prompt": "GREENROOF-II\n<br/><br/>\n5 points\n<br/>\nThe plants get the light energy from the sun and transform the energy to a new kind of energy called chemical energy. Also carbon dioxide, water, and the light energy come together and make the plants food, a sugar named glucose. They say that glucose is so energy rich because the light energy that was turned into chemical energy can be stored int the glucose for later and thats why these people have plants growing on there roof.\n<br/><br/>\n4 points\n<br/>\nThe energy comes from the sun and the energy transfers from place to place by bouncing off one another. The energy goes into the plant and is stored and is stored as chemical energy. Energy changes from light energy to chemical energy by the other energy's together..\n<br/><br/>\n3 points\n<br/>\nenergy comes from pre-existing energy that transforms, energy is stored in molecules. Energy explodes and the chemicals inside change.\n<br/><br/>\n2 points\n<br/>\nit would lower the energy use because it goes t the plants before it gets to the building\n<br/><br/>\n1 point\n<br/>\nhello world\n\n<img src=\"are_you_squidward.png\"/>Hi Lo",
    "showSaveButton": true,
    "showSubmitButton": true,
    "starterSentence": null,
    "isStudentAttachmentEnabled": true,
    "maxScore": 5,
    "enableNotifications": true,
    "notificationSettings": {
        "notifications": [
            {
                "notificationType": "CRaterResult",
                "enableCriteria": {
                    "score": [
                        1
                    ],
                    "scoreSequence": {
                        "0": "1",
                        "1": "1"
                    }
                },
                "isAmbient": true,
                "dismissCode": "apple",
                "isNotifyTeacher": true,
                "isNotifyStudent": true,
                "notificationMessageToStudent": "1. Please talk to your teacher about this CRater item!",
                "notificationMessageToTeacher": "1. {{username}} received score of {{score}}. Requires dismissal with code {{dismissCode}}."
            },
            {
                "notificationType": "CRaterResult",
                "enableCriteria": {
                    "score": [
                        2
                    ],
                    "scoreSequence": {
                        "0": "2",
                        "1": "2"
                    }
                },
                "isAmbient": true,
                "dismissCode": "apple",
                "isNotifyTeacher": true,
                "isNotifyStudent": true,
                "notificationMessageToStudent": "2. Please talk to your teacher about this CRater item!",
                "notificationMessageToTeacher": "2. {{username}} received score of {{score}}. Requires dismissal with code {{dismissCode}}."
            },
            {
                "notificationType": "CRaterResult",
                "enableCriteria": {
                    "score": [
                        3
                    ],
                    "scoreSequence": {
                        "0": "3",
                        "1": "3"
                    }
                },
                "isAmbient": true,
                "isNotifyTeacher": true,
                "isNotifyStudent": true,
                "notificationMessageToStudent": "3. Please talk to your teacher about this CRater item!",
                "notificationMessageToTeacher": "3. {{username}} received score of {{score}}. Requires dismissal with code {{dismissCode}}."
            }
        ]
    },
    "enableGlobalAnnotations": true,
    "globalAnnotationSettings": {
        "globalAnnotationMaxCount": 2,
        "globalAnnotationGroups": [
            {
                "annotationGroupName": "score1-2",
                "enableCriteria": {
                    "score": [
                        1,
                        2
                    ],
                    "scoreSequence": [
                        1,
                        2
                    ]
                },
                "maxCount": 1,
                "isPopup": true,
                "isGlobal": true,
                "unGlobalizeConditional": "all",
                "unGlobalizeCriteria": [
                    {
                        "name": "isVisitedAndRevisedAfter",
                        "params": {
                            "isVisitedAfterNodeId": "node2",
                            "isRevisedAfterNodeId": "node202",
                            "isRevisedAfterComponentId": "znco1zpt4n"
                        }
                    }
                ]
            },
            {
                "annotationGroupName": "score3",
                "enableCriteria": {
                    "score": [
                        3
                    ],
                    "scoreSequence": [
                        3,
                        3
                    ]
                },
                "maxCount": 1,
                "isPopup": true,
                "isGlobal": true,
                "unGlobalizeConditional": "all",
                "unGlobalizeCriteria": [
                    {
                        "name": "isVisitedAndRevisedAfter",
                        "params": {
                            "isVisitedAfterNodeId": "node2",
                            "isRevisedAfterNodeId": "node202",
                            "isRevisedAfterComponentId": "znco1zpt4n"
                        }
                    }
                ]
            },
            {
                "annotationGroupName": "score4-5",
                "enableCriteria": {
                    "score": [
                        4,
                        5
                    ],
                    "scoreSequence": [
                        4,
                        5
                    ]
                },
                "maxCount": 1,
                "isPopup": true,
                "isGlobal": true,
                "unGlobalizeConditional": "all",
                "unGlobalizeCriteria": [
                    {
                        "name": "isVisitedAfter",
                        "params": {
                            "isVisitedAfterNodeId": "node2"
                        }
                    }
                ]
            }
        ]
    },
    "cRater": {
        "itemType": "HENRY",
        "itemId": "GREENROOF-II",
        "scoreOn": "submit",
        "showScore": true,
        "showFeedback": true,
        "scoringRules": [
            {
                "concepts": "1",
                "score": "1",
                "feedbackText": "You got a score of 1. {{studentFirstNames}}, redo.<br><br> Think about the Sun’s rays. Where do they go? How do they move? Check out <wiselink node-id='node14' link-text='table step'></wiselink> for a hint. Then come back to <wiselink node-id='node46' link-text='this CRater step'></wiselink> and rewrite your story.",
                "studentAction": "revise"
            },
            {
                "concepts": "2",
                "score": "2",
                "feedbackText": "You got a score of 2. {{studentFirstNames}}, redo.<br><br> Think about the Sun’s rays. Where do they go? How do they move? Check out <wiselink node-id='node14' link-text='table step'></wiselink> for a hint. Then come back to <wiselink node-id='node46' link-text='this CRater step'></wiselink> and rewrite your story.",
                "studentAction": "revise"
            },
            {
                "concepts": "3",
                "score": "3",
                "feedbackText": "You got a score of 3. {{studentFirstNames}}, redo.<br><br> Think about the Sun’s rays. Where do they go? How do they move? Check out <wiselink node-id='node15' link-text='graph step'></wiselink> for a hint. Then come back to <wiselink node-id='node46' link-text='this CRater step'></wiselink> and rewrite your story.",
                "studentAction": "revise"
            },
            {
                "concepts": "4",
                "score": "4",
                "feedbackText": "You got a score of 4. {{studentFirstNames}}, redo.<br><br> Think about the Sun’s rays. Where do they go? How do they move? Check out <wiselink node-id='node15' link-text='graph step'></wiselink> for a hint",
                "studentAction": "revise"
            },
            {
                "concepts": "5",
                "score": "5",
                "feedbackText": "You got a score of 5. {{studentFirstNames}}, redo.<br><br> Think about the Sun’s rays. Where do they go? How do they move? Check out <wiselink node-id='node15' link-text='graph step'></wiselink> for a hint",
                "studentAction": "revise"
            },
            {
                "concepts": "",
                "score": "0",
                "feedbackText": "<br>{{studentFirstNames}}, expand your story.<br><br>How does the energy change inside the chloroplast? Check out <a onclick=\"view.goToStep('3.2')\">here</a> for a hint. Then, revise your story below.",
                "studentAction": "revise"
            }
        ]
    },
    "showAddToNotebookButton": true,
    "enableCRater": true,
    "rubric": "<p>Open Response Rubric22</p>\n<figure class=\"image\"><img src=\"av-3.png\" /></figure>\n<p>2</p>",
    "connectedComponents": [],
    "excludeFromTotalScore": false,
    "maxSubmitCount": null,
    "isStudentAudioRecordingEnabled": true
}
```

Closes #520

geoffreykwan committed:
chore(Notification): Convert dismiss ambient notification dialog

#520

hirokiterashima reviewed (changes requested):
Functionality looks good. 

I have few thoughts:
1. Can you add Notification type to the function params?
2. Instead of passing in the notificationService.dismissNotification() function to the DismissAmbientNotificationsDialog via data, would it be better to add an @Output to the DismissAmbientNotificationsDialog instead that emits an event that notificationService listens to. The idea is to remove the dismiss logic from the Dialog since it shouldn't need to know to do this. See this post: https://stackoverflow.com/a/61183011/4970939. I haven't tried it yet, but it looks like a cleaner solution.

geoffreykwan committed:
refactor(Notification): Call dismissNotification() directly

In the dismiss ambient notification dialog we were passing the function
into the dialog but now we are not passing it in and just calling it
directly using the NotificationService reference.

Also add Notification type to notification variables.

#520

geoffreykwan committed:
fix(Notification): Use observable instead of calling function directly

Calling the dismissNotification() function directly resulted in a
circular dependency. We've changed to use an observable that the dialog
opener can subscribe to.

#520

geoffreykwan committed:
refactor(Notification): Add Notification type to dialog data field

#520

hirokiterashima committed:
Pass in Notification to Dialog instead of wrapped object

hirokiterashima committed:
Add private function signatures
Refactor tests

hirokiterashima committed:
Refactor tests

#520
